src/sql/workbench/services/executionPlan/common/executionPlanService.ts

```diff
@@ -27,21 +27,59 @@ export class ExecutionPlanService implements IExecutionPlanService {
 		this._providerRegisterEvent = this._onProviderRegister.event;
 	}
 
-	/**
-	 * Runs the actions using the provider that supports the fileFormat provided.
-	 * @param fileFormat fileformat of the underlying execution plan file. It is used to get the provider that support it.
-	 * @param action executionPlanService action to be performed.
-	 */
-	private async _runAction<T>(fileFormat: string, action: (handler: azdata.executionPlan.ExecutionPlanProvider) => Thenable<T>): Promise<T> {
+	private async ensureCapabilitiesRegistered(): Promise<void> {
 		let providers = Object.keys(this._capabilitiesService.providers);
 		if (!providers) {
-			providers = await new Promise(resolve => {
+			await new Promise<void>(resolve => {
 				this._capabilitiesService.onCapabilitiesRegistered(e => {
-					resolve(Object.keys(this._capabilitiesService.providers));
+					resolve();
 				});
 			});
 		}
+	}
+
+	private async getExecutionPlanProvider(providerId: string): Promise<azdata.executionPlan.ExecutionPlanProvider> {
+		await this.ensureCapabilitiesRegistered();
+		const provider = this._capabilitiesService.providers[providerId];
+		// Return undefined if the provider is not registered or it is not a execution plan provider.
+		if (!provider || !provider.connection?.isExecutionPlanProvider) {
+			return undefined;
+		}
+
+		let handler = this._providers[providerId];
+		if (handler) {
+			return handler;
+		}
+		// Ensure the extension is loaded
+		await this._extensionService.whenInstalledExtensionsRegistered();
+		handler = this._providers[providerId];
+		if (!handler) {
+			handler = await new Promise((resolve, reject) => {
+				this._providerRegisterEvent(e => {
+					if (e.id === providerId) {
+						resolve(e.provider);
+					}
+				});
+				setTimeout(() => {
+					/**
+					 * Handling a possible edge case where provider registered event
+					 * might have been called before we await for it.
+					 */
+					resolve(this._providers[providerId]);
+				}, 30000);
```

Comment:
30 seconds is a long time to wait...why not instead just : 

1. Always await the promise (get rid of the initial providers map check)
2. Within the promise do the check in the map first and if it exists then immediately resolve
3. Still have a timeout but just return undefined (or reject) if it gets hit

This way in the edge case it'll still return immediately instead of blocking everything waiting for 30seconds unnecessarily